src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs

```diff
@@ -704,5 +707,52 @@ internal ITextBuffer GetOrCreateBufferForPath(string? filePath, IContentType con
                 return textBuffer;
             });
         }
+        internal override bool TryApplyChanges(
+            Solution newSolution,
+            IProgressTracker progressTracker)
+        {
+            var oldSolution = CurrentSolution;
+            if (base.TryApplyChanges(newSolution, progressTracker))
+            {
+                var projectChanges = newSolution.GetChanges(oldSolution).GetProjectChanges().ToList();
+                var changedDocs = projectChanges.SelectMany(pd => pd.GetChangedDocuments(true).Concat(pd.GetChangedAdditionalDocuments())).ToList();
+
+                NotifyRefactorChanges(changedDocs, newSolution, oldSolution);
```

Comment:
ick... i really don't trust thsi.  isn't this basically just testing behavior defined at the *tesT* layer?

Reply:
It is. I did my best to abstract a lot of the complex logic into `Rename.RenameSymbolAnnotation.GatherChangedSymbolsInDocumentsAsync` so it was as close as possible. I still need to add integration tests, which will E2E test the VS layer implementation. This does allow information to be asserted in unit tests though, so I would like something similar to this available.